src/Workspaces/Remote/Core/Services/Options/FileWatcher.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.IO;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    [Export(typeof(IFileWatcher)), Shared]
+    internal sealed class FileWatcher : IFileWatcher
+    {
+        private ImmutableDictionary<(string fileName, string directoryPath), FileSystemWatcher> _watchers = ImmutableDictionary<(string fileName, string directoryPath), FileSystemWatcher>.Empty;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public FileWatcher()
+        {
+        }
+
+        public event ConventionsFileChangedAsyncEventHandler ConventionFileChanged;
+        public event ContextFileMovedAsyncEventHandler ContextFileMoved;
+
+        public void Dispose()
+        {
+            var watchers = Interlocked.Exchange(ref _watchers, null);
+            if (watchers == null)
+            {
+                return;
+            }
+
+            ConventionFileChanged = null;
+            ContextFileMoved = null;
+        }
+
+        public void StartWatching(string fileName, string directoryPath)
+        {
+            FileSystemWatcher watcher = null;
+            var updated = ImmutableInterlocked.Update(
+                ref _watchers,
+                watchers =>
+                {
+                    if (!watchers.TryGetValue((fileName, directoryPath), out var existingWatcher))
+                    {
+                        if (watcher == null)
+                        {
+                            watcher = new FileSystemWatcher(directoryPath, fileName);
```

Comment:
are we sure that file system watcher per file has no perf issue?

Reply:
➡️ We are sure that it does have perf issues, especially when running on large numbers of files. Our primary redeeming factors here are:

1. This is only used OOP
2. This is only used as a workaround for not having hooked up the replacement .editorconfig support in OOP
3. This is only used for watching for .editorconfig files, and only one watcher is created for each directory

Reply:
> This is only used for watching for .editorconfig files

From a performance perspective, the kernel only implements directory watching: "file watching" is implemented by the framework by throwing away things not matching your filter. Not sure 3 is redeeming in this case (but 1 and 2 are "good enough" in my book.)